src/main/java/org/spongepowered/common/service/sql/SqlServiceImpl.java

```diff
@@ -190,9 +193,10 @@ public void close() throws IOException {
          * @param authlessUrl A JDBC url for this driver not containing authentication information
          * @param fullUrl The full jdbc url containing user, password, and database info
          */
-        public ConnectionInfo(@Nullable String user, @Nullable String password, String driverClassName, String authlessUrl, String fullUrl) {
+        public ConnectionInfo(@Nullable String user, @Nullable String password, String[] options, String driverClassName, String authlessUrl, String fullUrl) {
             this.user = user;
```

Comment:
Changing signature of public method leads to change of MAJOR version due to semver. You need to introduce new constructor if you do not want affect version number of the whole project.

Also you forgot to change `equals` method.

Reply:
That's ok, this is implementation change, not API.

Reply:
Anyway, this is breaking change. You can't be sure that no one using this class as it has public visibility.

Also I think, that introducing new field is unnecessary. My explanations about this, listed in another [review comment](https://github.com/SpongePowered/SpongeCommon/pull/2152#discussion_r238593182).

Reply:
You can simply make it the change unbreaking by passing the options as last parameter and using syntax for varargs instead of []

Reply:
WIP, it’ll be adjusted again tm.

Reply:
> Anyway, this is breaking change.

> You can't be sure that no one using this class as it has public visibility.

It doesn't matter here in SpongeCommon. We don't make _any_ guarantees on the implementation. Anyone doing so is doing so with the understanding that this might break build to build and is therefore at their own risk. We only guarantee that the Sponge API (and the associated dependencies with that) will not have any API breaks.

As long as it doesn't break API compatibility, do as you need to here.

Reply:
Public and protected modifiers always means that they are part of public exported api all over the world, but not in SpongeCommon :-/

Why not make the absence of guarantees explicit by making this class package-private? This make intentions more clear and less confusing for newcomers like me.

Reply:
If we were to do that, that would mean making every class in SpongeCommon package-private. However, since we use many different packages, that would mean SpongeCommon wouldn't be able to talk to itself.

The reason this isn't part of the API is that we already have an explicit API - SpongeAPI. Anything that isn't part of SpongeAPI is not part of the API. People can use the internal code if they want to, but we don't provide any expectations of compatibility.

(Most of the major Minecraft projects do it the same way. Spigot has lots of public classes, but it explicitly specifies which ones are part of its API, so everything else is not part of the API. Same thing with BungeeCord, Waterfall, PaperSpigot, Bukkit/CraftBukkit, Canary, and others.)

Reply:
Yeah, a lot of refactoring and project architecture revision required.

I'm upset, that all the projects you listed ignores such common and widespread programming practices.

Reply:
Assuming that public methods are "public" in the sense that developers are expected to interact with them is unfair, really. To design a project perfectly such that this is the case seems unlikely. Often, non-static methods from one package may need to be used in another (non-sub) package.

Reply:
It's not hard to make good design with adequate encapsulation from the ground up (look at the java standard library, for example). It is much more difficult to put in order the design of an already formed project. It's like putting a skeleton in meat. And finally, it is truly impossible to convince the community of the usefulness and necessity of such fundamental changes.

Reply:
It's definitely sensible, but often you'll have a catch-all class or package for handlers, registration of sorts, etc. and multiple packages will hook into those "nodes." Of course, it's possible to implement an abstract class of these types when you need one, but you won't find much of that in Minecraftia.

Reply:
I don't particularly want to silence this conversation, because I think it's important that we take the time to discuss things with others and explain our thinking if people would like to do so. That said, this is off topic for this PR now, so I'd like to ask that futher conversation regarding this, if any is needed, including asking about why we chose this direction, is directed to our forums or other channels, so we can focus on getting this PR reviewed. 

Thanks! :)

Reply:
The method signature break is fine. Anyone who builds against the implementation does so under their own peril.